// @ephox/swag/lib/main/ts/ast/PatchExports.js
"use strict";
exports.__esModule = true;
var Uuid = require("../utils/Uuid");
var patch = function (node) {
    var id = Uuid.generate('$');
    node.body = node.body.reduce(function (acc, node) {
        if (node.type === 'ExportDefaultDeclaration') {
            var declaration = node.declaration;
            if (declaration.type === 'ObjectExpression') {
                return acc.concat([
                    {
                        type: 'VariableDeclaration',
                        declarations: [
                            {
                                type: 'VariableDeclarator',
                                id: {
                                    type: 'Identifier',
                                    name: id
                                },
                                init: declaration
                            }
                        ],
                        kind: 'var'
                    },
                    {
                        type: 'ExportDefaultDeclaration',
                        declaration: {
                            type: 'Identifier',
                            name: id
                        }
                    }
                ]);
            }
        }
        return acc.concat([node]);
    }, []);
};
exports.patch = patch;
