// @ephox/alloy/lib/main/ts/ephox/alloy/api/events/NativeEvents.js
import { Fun } from '@ephox/katamari';
var contextmenu = Fun.constant('contextmenu');
var touchstart = Fun.constant('touchstart');
var touchmove = Fun.constant('touchmove');
var touchend = Fun.constant('touchend');
var gesturestart = Fun.constant('gesturestart');
var mousedown = Fun.constant('mousedown');
var mousemove = Fun.constant('mousemove');
var mouseout = Fun.constant('mouseout');
var mouseup = Fun.constant('mouseup');
var mouseover = Fun.constant('mouseover');
// Not really a native event as it has to be simulated
var focusin = Fun.constant('focusin');
var keydown = Fun.constant('keydown');
var input = Fun.constant('input');
var change = Fun.constant('change');
var focus = Fun.constant('focus');
var click = Fun.constant('click');
var transitionend = Fun.constant('transitionend');
var selectstart = Fun.constant('selectstart');
export { contextmenu, touchstart, touchmove, touchend, gesturestart, mousedown, mousemove, mouseout, mouseup, mouseover, focusin, keydown, input, change, focus, click, transitionend, selectstart, };
