// @ephox/swag/lib/main/ts/api/NodeResolve.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
exports.__esModule = true;
var Resolve_1 = require("../fs/Resolve");
var CachedFileSystem_1 = require("../fs/CachedFileSystem");
var defaultOptions = {
    basedir: '.',
    prefixes: [],
    forceFlat: true
};
var nodeResolve = function (options) {
    if (options === void 0) { options = defaultOptions; }
    var combinedOptions = __assign({}, defaultOptions, options);
    var prefixes = Resolve_1.resolvePrefixPaths(combinedOptions.basedir, combinedOptions.prefixes);
    var fs = options.fileSystem ? options.fileSystem : CachedFileSystem_1.getFileSystem();
    return {
        name: 'swag-resolve',
        resolveId: Resolve_1.resolveId(fs, prefixes, combinedOptions.forceFlat)
    };
};
exports.nodeResolve = nodeResolve;
