// @ephox/alloy/lib/main/ts/ephox/alloy/ui/schema/TabSectionSchema.js
import { FieldSchema } from '@ephox/boulder';
import { Fun } from '@ephox/katamari';
import * as SketchBehaviours from '../../api/component/SketchBehaviours';
import { Tabbar } from '../../api/ui/Tabbar';
import { Tabview } from '../../api/ui/Tabview';
import * as Fields from '../../data/Fields';
import * as PartType from '../../parts/PartType';
var schema = Fun.constant([
    FieldSchema.defaulted('selectFirst', true),
    Fields.onHandler('onChangeTab'),
    Fields.onHandler('onDismissTab'),
    FieldSchema.defaulted('tabs', []),
    SketchBehaviours.field('tabSectionBehaviours', [])
]);
var barPart = PartType.required({
    factory: Tabbar,
    schema: [
        FieldSchema.strict('dom'),
        FieldSchema.strictObjOf('markers', [
            FieldSchema.strict('tabClass'),
            FieldSchema.strict('selectedClass')
        ])
    ],
    name: 'tabbar',
    defaults: function (detail) {
        return {
            tabs: detail.tabs()
        };
    }
});
var viewPart = PartType.required({
    factory: Tabview,
    name: 'tabview'
});
var parts = Fun.constant([
    barPart,
    viewPart
]);
var name = Fun.constant('TabSection');
export { name, schema, parts };
