// @ephox/alloy/lib/main/ts/ephox/alloy/api/ui/Container.js
import * as SketchBehaviours from '../component/SketchBehaviours';
import * as Sketcher from './Sketcher';
import { FieldSchema } from '@ephox/boulder';
import { Merger } from '@ephox/katamari';
var factory = function (detail, spec) {
    return {
        uid: detail.uid(),
        dom: Merger.deepMerge({
            tag: 'div',
            attributes: {
                role: 'presentation'
            }
        }, detail.dom()),
        components: detail.components(),
        behaviours: SketchBehaviours.get(detail.containerBehaviours()),
        events: detail.events(),
        domModification: detail.domModification(),
        eventOrder: detail.eventOrder()
    };
};
var Container = Sketcher.single({
    name: 'Container',
    factory: factory,
    configFields: [
        FieldSchema.defaulted('components', []),
        SketchBehaviours.field('containerBehaviours', []),
        // TODO: Deprecate
        FieldSchema.defaulted('events', {}),
        FieldSchema.defaulted('domModification', {}),
        FieldSchema.defaulted('eventOrder', {})
    ]
});
export { Container };
