// @ephox/alloy/lib/main/ts/ephox/alloy/api/ui/ModalDialog.js
import { Merger } from '@ephox/katamari';
import { Traverse } from '@ephox/sugar';
import * as AlloyParts from '../../parts/AlloyParts';
import * as ModalDialogSchema from '../../ui/schema/ModalDialogSchema';
import * as Behaviour from '../behaviour/Behaviour';
import { Keying } from '../behaviour/Keying';
import * as GuiFactory from '../component/GuiFactory';
import * as SketchBehaviours from '../component/SketchBehaviours';
import * as Attachment from '../system/Attachment';
import * as Sketcher from './Sketcher';
var factory = function (detail, components, spec, externals) {
    // TODO IMPROVEMENT: Make close actually close the dialog by default!
    var showDialog = function (dialog) {
        var sink = detail.lazySink()().getOrDie();
        var blocker = sink.getSystem().build(Merger.deepMerge(externals.blocker(), {
            components: [
                GuiFactory.premade(dialog)
            ]
        }));
        Attachment.attach(sink, blocker);
        Keying.focusIn(dialog);
    };
    var hideDialog = function (dialog) {
        Traverse.parent(dialog.element()).each(function (blockerDom) {
            dialog.getSystem().getByDom(blockerDom).each(function (blocker) {
                Attachment.detach(blocker);
            });
        });
    };
    var getDialogBody = function (dialog) {
        return AlloyParts.getPartOrDie(dialog, detail, 'body');
    };
    return {
        dom: Merger.deepMerge({
            attributes: {
                role: 'dialog'
            }
        }, detail.dom()),
        components: components,
        apis: {
            show: showDialog,
            hide: hideDialog,
            getBody: getDialogBody
        },
        behaviours: Merger.deepMerge(Behaviour.derive([
            Keying.config({
                mode: 'cyclic',
                onEnter: detail.onExecute(),
                onEscape: detail.onEscape(),
                useTabstopAt: detail.useTabstopAt()
            })
        ]), SketchBehaviours.get(detail.modalBehaviours()))
    };
};
var ModalDialog = Sketcher.composite({
    name: 'ModalDialog',
    configFields: ModalDialogSchema.schema(),
    partFields: ModalDialogSchema.parts(),
    factory: factory,
    apis: {
        show: function (apis, dialog) {
            apis.show(dialog);
        },
        hide: function (apis, dialog) {
            apis.hide(dialog);
        },
        getBody: function (apis, dialog) {
            return apis.getBody(dialog);
        }
    }
});
export { ModalDialog };
